src/libproc_macro/lib.rs

```diff
@@ -200,6 +201,78 @@ impl Span {
     pub fn call_site() -> Span {
         ::__internal::with_sess(|(_, mark)| Span(mark.expn_info().unwrap().call_site))
     }
+
+    /// The original source file into which this span points.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn source_file(&self) -> Option<SourceFile> {
+        // If this span's token was produced by a macro
+        if self.0.ctxt != SyntaxContext::empty() {
+            return None
+        }
+
+        let loc = __internal::lookup_char_pos(self.0.lo);
+
+        Some(SourceFile(loc.file))
+    }
+
+    /// Get the starting line/column in the source file for this span.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn start(&self) -> LineColumn {
+        let loc = __internal::lookup_char_pos(self.0.lo);
+        LineColumn {
+            line: loc.line,
+            column: loc.col.to_usize()
+        }
+    }
+
+    /// Get the ending line/column in the source file for this span.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn end(&self) -> LineColumn {
+        let loc = __internal::lookup_char_pos(self.0.hi);
+        LineColumn {
+            line: loc.line,
+            column: loc.col.to_usize()
+        }
+    }
+
+    /// Create a new span encompassing `self` and `other`.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn glue(&self, other: Span) -> Span {
```

Comment:
Is `glue()` a name already used today? Maybe `hull()` is more fitting?

Reply:
Oh and regarding your `glue_many()` idea: I wouldn't make it a method, but an associated function:

```rust
fn hull_of(spans: &[Span]) -> Span
```

Reply:
`hull` is a bit arcane for terminology. `syntax_pos`'s `Span` uses `to()` but I feel like that's too short and undescriptive for a public API. Maybe `to_other()`?

Reply:
I think the term [*convex hull*](https://en.wikipedia.org/wiki/Convex_hull) is actually the mathematically correct one here (although usually, 1D  convex hulls are not that interesting). So `hull()` makes sense and is short. I'm also fine with `Span::containing()`, but it's longer than `hull()`. I don't like `to()` or `to_other()`...

Reply:
It's too much to explain in a short piece of documentation. We're looking for something that immediately makes sense.

Reply:
`join()` isn't terrible, though.